### PR TITLE
first draft

### PR DESCRIPTION
**Pull request by Landay7**

### CODE REVIEW

File: src/helpers/zip_helper.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Helper code, might be used in the future
+
+use std::io::prelude::*;
+
+fn get_file_as_byte_vec(filename: &String) -> Vec<u8> {
```
Comment (vasyl-protsiv) [inline src/helpers/zip_helper.rs]: this is called `std::fs::read`

File: src/layers/activation_layer.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                softmax_activation(incoming)
+            },
+            ActivationFunction::Linear => {
+                incoming
+            }
+        }
+    }
+}
+
+impl Into<ActivationFunction> for String{
```
Comment (vasyl-protsiv) [inline src/layers/activation_layer.rs]: It is generally better to implement `From` trait instead of `Into`, since `From` allows both `ActivationFunction::from(string)` and `string.into::<ActivationFunction>()`.
Also consider using `&str` instead of `String`, as this would allow using `from` without consuming string.

File: src/layers/activation_layer.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    fn into(self) -> ActivationFunction {
+        match self.as_str() {
+            "relu" => ActivationFunction::ReLu,
+            "sigmoid" => ActivationFunction::Sigmoid,
+            "softmax" => ActivationFunction::SoftMax,
+            "linear" => ActivationFunction::Linear,
+            _ => panic!("Unimplemented Activation function {}", self)
```
Comment (vasyl-protsiv) [inline src/layers/activation_layer.rs]: Since this can fail consider implementing `TryFrom` instead

File: src/model/sequantial.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+impl SequentialModel {
+    pub fn from_config_and_hdf5(config: Config, file: &hdf5::File) -> Self {
+        let mut layers = Vec::new();
+        for layer_config in config.config.layers {
+            let layer = match layer_config.class_name.as_str() {
+                "Dense" => {
+                    let activation = layer_config.config["activation"].as_str().map(|val| {
+                        val.to_string().into()
```
Comment (vasyl-protsiv) [inline src/model/sequantial.rs]: if you derive `Deserialize` for `ActivationFunction` you should be able to replace this with
```rust
let activation = serde_json::from_value(layer_config.config["activation"]);
```
or maybe parse all relevant values in config with serde

File: src/model/sequantial.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+pub struct SequentialModel {
+    layers: Vec<Box<dyn Layer>>,
+}
+
+impl SequentialModel {
+    pub fn from_config_and_hdf5(config: Config, file: &hdf5::File) -> Self {
+        let mut layers = Vec::new();
+        for layer_config in config.config.layers {
+            let layer = match layer_config.class_name.as_str() {
```
Comment (vasyl-protsiv) [inline src/model/sequantial.rs]: you can set explicit type for layer
```rust
let layer: Box<dyn Layer> = ...
```
then you can remove as `as Box<dyn Layer>` in each case.

File: src/model/sequantial.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                        val.to_string().into()
+                    });
+                    Box::new(Dense::from_hdf5(file, layer_config.config["name"].as_str().unwrap(), activation)) as Box<dyn Layer>
+                }
+                "Flatten" => {
+                    Box::new(Flatten) as Box<dyn Layer>
+                }
+                "InputLayer" => { continue; }
+                _ => {
+                    panic!("Unsupported layer type: {}", layer_config.class_name);
```
Comment (vasyl-protsiv) [inline src/model/sequantial.rs]: consider returning result instead

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    max_index
+}
+
+fn main() {
+    let hdf5_file = hdf5::File::open("model.weights.h5").expect("Failed to open HDF5 file");
+    let deserialized: Config = serde_json::from_str(fs::read_to_string("config.json").unwrap().as_str()).unwrap();
+    let model = SequentialModel::from_config_and_hdf5(deserialized, &hdf5_file);
+    let mut npz = NpzReader::new(File::open("test_data.npz").unwrap()).unwrap();
+    let x: Array4<f32> = npz.by_name("X.npy").unwrap();
+    let y: Array1<i64> = npz.by_name("y.npy").unwrap();
```
Comment (vasyl-protsiv) [inline src/main.rs]: Consider using result and ? operator instead of unwrap

**Review by vasyl-protsiv — COMMENTED** [review]
Error handling can be improved by using results instead of panics in many places, particularly in dataset parsing.
There are few places where you rely on strings, that could probably be replaced with enums or other more type safe alternatives.
Maybe mention that hdf5 should be installed on system in order to build project.
Also mention where to get `model.weights.h5`, `config.json` and other necessary files.

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+.DS_Store
+
+/.ipynb_checkpoints/
+/target/
+*.npz
+*.h5
+
+config.json
+metadata.json
+*.keras
```
Comment (tyranron) [inline .gitignore]: No trailing newline is a bad tone!

File: src/helpers/zip_helper.rs (hunk 1 above)
Comment (Landay7) [reply src/helpers/zip_helper.rs]: Thank you, but for now I don't need this function, so I delete it.

File: src/layers/activation_layer.rs (hunk 2 above)
Comment (Landay7) [reply src/layers/activation_layer.rs]: I follow you suggestion and use just `serde_json::from_value`

File: src/layers/activation_layer.rs (hunk 3 above)
Comment (Landay7) [reply src/layers/activation_layer.rs]: fixed

File: src/model/sequantial.rs (hunk 5 above)
Comment (Landay7) [reply src/model/sequantial.rs]: thank you, fixed

File: src/model/sequantial.rs (hunk 6 above)
Comment (Landay7) [reply src/model/sequantial.rs]: fixed

File: src/main.rs (hunk 7 above)
Comment (Landay7) [reply src/main.rs]: fixed. I also add custom Error :)

**Review by vasyl-protsiv — APPROVED** [review]

**Comment by NamesMark** [comment]
🫡 good code

File: README.md
Comment (oleksandrsan) [inline README.md:1]: Change the GIT attribute of this file to text.
To automate file extraction, you can also add the command `! unzip model.keras` in the Jupyter Notebook.

File: src/activations/functions.rs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+const E: f32 = std::f32::consts::E;
+
+pub fn softmax_activation(z: NArray) -> NArray {
+    let exp_arr = z.mapv(|x| f32::powf(E, x));
+    let sum: f32 = exp_arr.iter().sum();
+    exp_arr.mapv(|x| x / sum)
+}
+
+pub fn sigmoid_activation(z: NArray) -> NArray {
+    z.mapv(|x| sigmoid(x))
```
Comment (oleksandrsan) [inline src/activations/functions.rs]: Minor: you can also write `z.mapv(sigmoid)`

File: src/main.rs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        .by_name("X.npy")
+        .map_err(MainError::NpzMissingFileError)?;
+    let y: Array1<i64> = npz
+        .by_name("y.npy")
+        .map_err(MainError::NpzMissingFileError)?;
+
+    let mut index: usize = 0;
+    let mut correct = 0;
+    let now = Instant::now();
+    for case in x.axis_iter(Axis(0)) {
```
Comment (oleksandrsan) [inline src/main.rs]: Minor: You can use `enumerate` instead of manually incrementing the index

**Review by oleksandrsan — APPROVED** [review]
Nice work .
You can move the main.rs file to the examples folder, so that it's easier to add different cases down the road.
Also, it will be interesting to see how your code compares to the Keras implementation.

File: README.md
Comment (tyranron) [inline README.md:1]: There are some instruction in the README, however, no clue about what the heck the project is.

File: Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "rust_deep_learning"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (tyranron) [inline Cargo.toml:6]: If the project is going to be published on crates.io, it would be good to add rich metadata and licensing. If not, then it's better to add `publish = false`, to prevent accidental publishes.

File: README.md
Comment (tyranron) [reply README.md:1]: And yes, for any open-source work adding a license is vital, as gives people understanding how they can use the published sources.

File: src/model/sequential.rs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    pub fn from_config_and_hdf5(config: Config, file: &hdf5::File) -> Result<Self, ModelError> {
+        let mut layers = Vec::new();
+        for layer_config in config.get_layers() {
+            let layer: Box<dyn Layer> = match layer_config.get_class_name() {
+                LayerType::Dense => {
+                    let activation =
+                        serde_json::from_value(layer_config.get_property("activation").clone())
+                            .map_err(ModelError::ParsingError)?;
+                    let layer_name = layer_config.get_property("name").as_str().ok_or(
+                        ModelError::ConfigurationError("Failed to find layer name".to_string()),
```
Comment (tyranron) [inline src/model/sequential.rs]: In such case, it's much better to use `.into()`. Whenever the type reasonably changes (like SSO `String` -> `SmartString` for example), you won't need to update the code here.

File: src/activations/mod.rs [hunk 13]
```diff
@@ -0,0 +1 @@
+pub mod functions;
```
Comment (tyranron) [inline src/activations/mod.rs]: Why do have `functions` module at all? Why not just put functions in the `activations` module directly. Is the fact the `fn` is a function already not enough?

File: src/activations/functions.rs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+pub fn relu_activation(z: NArray) -> NArray {
```
Comment (tyranron) [inline src/activations/functions.rs]: And again, why use `_activation` postfix everywhere?

Instead of `activations::functions::relu_activation()` you could just have `activation::relu()`.

File: src/activations/functions.rs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+#[cfg(test)]
+mod tests {
+    use crate::Vector;
+
+    use super::*;
+    use assert_approx_eq::assert_approx_eq;
+
+    #[test]
+    fn test_softmax_activation() {
```
Comment (tyranron) [inline src/activations/functions.rs]: `test_` prefix here similarly seems to be redundant. Why read `tests::test_softmax_activation` in tests output rather than `test::softmax_activation`?

Overall, it's better for the test name to reflect the behavior being tested, not the thing being tested.

File: src/layers/dense.rs [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+            activation,
+        }
+    }
+
+    pub fn from_hdf5(
+        file: &hdf5::File,
+        layer_name: &str,
+        activation: Option<ActivationFunction>,
+    ) -> Result<Self, hdf5::Error> {
+        let base_path = format!(r"/layers\{}/vars", layer_name);
```
Comment (tyranron) [inline src/layers/dense.rs]: Why not `format!(r"/layers\{layer_name}/vars")`?

We even have a [lint for this](https://rust-lang.github.io/rust-clippy/master/index.html#/uninlined_format_args).

File: src/layers/activation_layer.rs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    #[serde(alias = "softmax")]
+    SoftMax,
+    #[serde(alias = "linear")]
+    Linear,
+}
+
+impl ActivationFunction {
+    pub fn compute(&self, incoming: NArray) -> NArray {
+        match self {
+            ActivationFunction::ReLu => relu_activation(incoming),
```
Comment (tyranron) [inline src/layers/activation_layer.rs]: It's also recommended to use `Self` as much as possible. Here, the match expression would be more ergonomic if written using `Self`:
```rust
match self {
    Self::ReLu => relu_activation(incoming),
    Self::Sigmoid => sigmoid_activation(incoming),
    Self::SoftMax => softmax_activation(incoming),
    Self::Linear => incoming,
}
```

File: src/layers/layer_trait.rs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+use crate::{Matrix, NArray, Vector};
+
+pub trait Layer {
```
Comment (tyranron) [inline src/layers/layer_trait.rs]: Why keep it in a separate module? As a general abstraction, it would be totally OK to keep in the `layer` module itself.

File: src/model/sequential.rs (hunk 12 above)
Comment (tyranron) [reply src/model/sequential.rs]: Even more, why keep `String` here at all? It's use only here, so `&'static str` is enough.

File: src/main.rs [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+use ndarray_npy::NpzReader;
+use rust_deep_learning::configuration::config::Config;
+use rust_deep_learning::model::sequential::SequentialModel;
+use rust_deep_learning::Vector;
+use std::fs::{self, File};
+use std::time::Instant;
+
+#[derive(Debug)]
+enum MainError {
+    Hdf5OpenError(String),
```
Comment (tyranron) [inline src/main.rs]: It's generally a good tone to implement `Display` and `Error` traits for the error type, rather than using `Debug` and imitating the `Display`ing by keeping `String`s instead of the original upstream error types.

File: src/main.rs [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        })?;
+
+        // Find the index of the maximum element (argmax)
+        let argmax_index = argmax(&array1).ok_or(MainError::ArrayTransformError(
+            "Failed to find argmax index".to_string(),
+        ))?;
+        if argmax_index as i64 == y[index] {
+            correct += 1;
+        }
+        index += 1;
```
Comment (tyranron) [inline src/main.rs]: Why use "by hand" indexing instead of 
```rust
for (index, case) in x.axis_iter(Axis(0)).enumerate() {
```
?

File: src/main.rs [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    let elapsed = now.elapsed();
+    println!("Elapsed: {:.2?}", elapsed);
+    println!("{}", correct);
+
+    Ok(())
+}
+
+fn argmax(array: &Vector) -> Option<usize> {
```
Comment (tyranron) [inline src/main.rs]: Why writing own function instead of using [`Itertools::position_max()`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.position_max)?

**Review by tyranron — COMMENTED** [review]
Overall, the major problem, I see with the project, is the total lack of the documentation.

Another ones, which catch the eye instantly, are:
1. Design of the error types. I would expect usage of `thiserror` or `derive_more` rather than `String`y variants here and there.
2. Modules layout. They're decomposed way too granular, which makes them harder to use and navigate. The "rusty way" is not to keep each definition in a separate file/module, but rather unite definitions under a single reasonable namespace as a logical unit/module.

**Comment by NamesMark** [comment]
After the Mr Kai's comments I don't have any :) Great project, looking forward to see the next iteration. :)

**Review by NamesMark — APPROVED** [review]
lgtm :)

File: README.md
Comment (Landay7) [reply README.md:1]: fixed

File: .gitignore (hunk 8 above)
Comment (Landay7) [reply .gitignore]: fixed

File: README.md
Comment (Landay7) [reply README.md:1]: I added license and tried to improve README.

File: Cargo.toml (hunk 11 above)
Comment (Landay7) [reply Cargo.toml:6]: fixed

File: src/activations/mod.rs (hunk 13 above)
Comment (Landay7) [reply src/activations/mod.rs]: Fixed. Now it has module activations.

File: src/activations/functions.rs (hunk 14 above)
Comment (Landay7) [reply src/activations/functions.rs]: fixed

File: src/activations/functions.rs (hunk 15 above)
Comment (Landay7) [reply src/activations/functions.rs]: fixed

File: src/layers/dense.rs (hunk 16 above)
Comment (Landay7) [reply src/layers/dense.rs]: My bad, fixed.

File: src/layers/activation_layer.rs (hunk 17 above)
Comment (Landay7) [reply src/layers/activation_layer.rs]: Fixed

File: src/layers/layer_trait.rs (hunk 18 above)
Comment (Landay7) [reply src/layers/layer_trait.rs]: I didn't know that it's okay to have some more information at `mod.rs` file. 
Fixed

File: src/main.rs (hunk 21 above)
Comment (Landay7) [reply src/main.rs]: I didn't know about that. Thank you, fixed.

File: src/main.rs (hunk 20 above)
Comment (Landay7) [reply src/main.rs]: Cool, rust also have enumerate. I didn't know about that.
Thank you

File: src/activations/functions.rs (hunk 9 above)
Comment (Landay7) [reply src/activations/functions.rs]: fixed

File: src/main.rs (hunk 10 above)
Comment (Landay7) [reply src/main.rs]: fixed

File: src/model/sequential.rs (hunk 12 above)
Comment (Landay7) [reply src/model/sequential.rs]: Fixed

File: src/main.rs (hunk 19 above)
Comment (Landay7) [reply src/main.rs]: Fixed

File: LICENSE [hunk 22]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2017 Matthew D. Steele
```
Comment (tyranron) [inline LICENSE]: 🙈

**Review by tyranron — CHANGES_REQUESTED** [review]

File: LICENSE (hunk 22 above)
Comment (Landay7) [reply LICENSE]: Lol, thank you!

**Review by tyranron — APPROVED** [review]

**Review by divi255 — APPROVED** [review]
